kite-data/kite-data-oozie/src/main/java/org/kitesdk/data/oozie/KiteURIHandler.java

```diff
@@ -0,0 +1,154 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kitesdk.data.oozie;
+
+import java.net.URI;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.oozie.ErrorCode;
+import org.apache.oozie.XException;
+import org.apache.oozie.action.hadoop.LauncherURIHandler;
+import org.apache.oozie.dependency.URIHandler;
+import org.apache.oozie.dependency.URIHandlerException;
+import org.apache.oozie.service.URIHandlerService;
+import org.kitesdk.data.DatasetNotFoundException;
+import org.kitesdk.data.Datasets;
+import org.kitesdk.data.Signalable;
+import org.kitesdk.data.URIBuilder;
+import org.kitesdk.data.View;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A Kite URI handler that works with {@link Signalable} views.
+ *
+ * To be considered as {@link #exists(URI, Configuration, String) existing} the
+ * view must have been signaled as ready.
+ */
+public class KiteURIHandler implements URIHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KiteURIHandler.class);
+
+  private Set<String> supportedSchemes;
+  private List<Class<?>> classesToShip;
+
+  @Override
+  public void init(final Configuration conf) {
+    supportedSchemes = new HashSet<String>();
+    final String[] schemes = conf.getStrings(
+        URIHandlerService.URI_HANDLER_SUPPORTED_SCHEMES_PREFIX
+            + this.getClass().getSimpleName()
+            + URIHandlerService.URI_HANDLER_SUPPORTED_SCHEMES_SUFFIX, "view", "dataset");
+    supportedSchemes.addAll(Arrays.asList(schemes));
+    classesToShip = new KiteLauncherURIHandler().getClassesForLauncher();
+  }
+
+  @Override
+  public Set<String> getSupportedSchemes() {
+    return supportedSchemes;
+  }
+
+  @Override
+  public Class<? extends LauncherURIHandler> getLauncherURIHandlerClass() {
+    return KiteLauncherURIHandler.class;
+  }
+
+  @Override
+  public List<Class<?>> getClassesForLauncher() {
+    return classesToShip;
+  }
+
+  @Override
+  public DependencyType getDependencyType(final URI uri)
+      throws URIHandlerException {
+    return DependencyType.PULL;
+  }
+
+  @Override
+  public void registerForNotification(final URI uri, final Configuration conf,
+      final String user, final String actionID) throws URIHandlerException {
+    throw new UnsupportedOperationException(
+        "Notifications are not supported for " + uri);
+  }
+
+  @Override
+  public boolean unregisterFromNotification(final URI uri, final String actionID) {
+    throw new UnsupportedOperationException(
+        "Notifications are not supported for " + uri);
+  }
+
+  @Override
+  public Context getContext(final URI uri, final Configuration conf,
```

Comment:
We probably should be implementing this method.  The method is a mechanism to optimize a series of calls to exist().  Its usage can be seen in [CoordELFucntions.coord_futureRange_sync](https://github.com/apache/oozie/blob/release-4.0.0/core/src/main/java/org/apache/oozie/coord/CoordELFunctions.java#L313) and [coord_latestRange_sync](https://github.com/apache/oozie/blob/release-4.0.0/core/src/main/java/org/apache/oozie/coord/CoordELFunctions.java#L1044).

The coord_latestRange_sync method is used to look for the [latest complete instance (or instances) of a dataset](https://oozie.apache.org/docs/4.0.0/CoordinatorFunctionalSpec.html#a6.6.6._coord:latestint_n_EL_Function_for_Synchronous_Datasets) when coord:future() is used in the coordinator definition.  Oozie will loop backwards from the latest nominal time all the way back to the initial instance of the dataset looking for an instance that "exists" (is complete).  In the worst case you have a dataset that has been around for along time and the latest instance is very old relative to the frequency of the dataset.  For example, if the dataset is declared with a frequency of once per day with initial instance declared as 1 year ago and the only instance that is complete is actually one year ago then latest(0) will have to loop through 365 URIs checking exists() before returning that instance.

The coord_futureRange_sync method supports [coord:future()](https://oozie.apache.org/docs/4.0.0/CoordinatorFunctionalSpec.html#a6.6.7._coord:futureint_n_int_limit_EL_Function_for_Synchronous_Datasets) and loops forward looking for the desired number of compete instances.

FSURIHandler [uses getContext() to cache the Filesystem instance](https://github.com/apache/oozie/blob/release-4.0.0/core/src/main/java/org/apache/oozie/dependency/FSURIHandler.java#L82).

HCatURIHandler [caches the HCatClient](https://github.com/apache/oozie/blob/release-4.0.0/core/src/main/java/org/apache/oozie/dependency/HCatURIHandler.java#L143).

I think we could cache the View instance so we don't have as heavy of a hit on each exists() check to load up all the Dataset metadata.

Separately from talking about this specific method, I had talked to @bbrownz about the idea that perhaps we might want to follow up later with a JIRA to introduce some other caching for optimization.  In an environment with lots of Oozie coordinators using Kite datasets and this URI handler, the Oozie server could be making very frequent calls to exists().  Having each call hit HDFS and/or the Hive metastore to load dataset metadata (on Datasets.load()) might take a toll that could be decreased with some more caching of the Dataset instances.

There is also the possibility that we could follow up later with an attempt to support a push model (DependencyType.PUSH) instead of a polling model for Hive backed datasets.

In any case, I think what is here works and is probably worth merging in as-is and following up with the optimizations in separate later JIRAs.  Perhaps we implement the getContext() piece now though as it is the simplest although it only comes into play in certain specific use cases (using coord:latest() or coord:future() - neither of which I expect my organization will actually use anytime soon as our use cases prefer coord:current()).